### DISCUSSION TITLE
Is there a way to read uncommitted changes within a transaction?

### DISCUSSION THREAD

**Discussion by sudall** [prisma/prisma#23370] [Q&A]
### Question

I've encountered this pseudocode situation:

```ts
async function removeFromGroup(userId: number) {
  const groupMember = await prisma.groupMember.findUnique({
    where: {
      userId: userId,
    },
    select: {
      id: true,
      isLeader: true,
      group: {
        select: {
          membersVersion: true,
        },
      },
    },
  })

  const actions = []
  const groupUpdateAction = prisma.group.update({
    where: {
      id: groupMember.group.id,
      membersVersion: groupMember.group.membersVersion,
    },
    data: {
      membersVersion: {
        increment: 1,
      },
    },
  })
  actions.push(groupUpdateAction)

  // delete the group member...
  actions.push(deleteGroupMemberAction)

  if (groupMemberCount === 1) {
    // delete the group
    actions.push(deleteGroupAction)
  } else {
    if (groupMember.isLeader) {
      // give another member leadership
      actions.push(transferLeadershipAction)
    }
  }

  return actions
}

async function doThingWhereINeedToRemoveUsersFromTheirGroup {
  const dbActions = []

  for (let user of users) {
    const removeActions = await removeFromGroup(user.id)
    dbActions.push(...removeActions)
  }

  await prisma.$transaction(dbActions)
}

```

The issue arises when removing two different users from the same group in the same transaction. Each call to `removeFromGroup()` is going to read the same group info via `groupMember.findUnique`. So, if I have two members in a group and I try to remove them both in `doThingWhereINeedToRemoveUsersFromTheirGroup()`, an error will occur because they try to make an update on the same `membersVersion`.

The way that I'm thinking this could be solved is if there's a way to read uncommited changes (as is possible with Read Committed isolation) so that sequential `groupMember.findUnique` will read the previous removeFromGroup() uncommited changes. I couldn't find a way to do this. I'm aware that interactive transactions can do it, but is there a way to compose interactive transactions together?

If this isn't possible right now, I think something like these ideas would be really nice:

```ts
const findGroupMemberAction = prisma.groupMember.findUnique(...)
// make the findUnique() part of the transaction so that it can read uncommitted changes
actions.push(findGroupMemberAction)

// and then be able to use the uncommitted changes somehow
// ... update group where `membersVersion` = `findGroupMemberAction.group.membersVersion`
```

or

```ts
const transactionActions = prisma.$transaction(() => {
   // remove group member
})

const transactionActions2 = prisma.$transaction(() => {
  // remove group member
})

// assuming the second transaction can read uncommitted changes from the first
await prisma.$transaction([transactionActions, transactionActions2, ...otherActions])
```

Any ideas or help is greatly appreciated.

**Comment by ludralph** [comment]
Hi @sudall 👋 

Thank you for raising this question.

Prisma does support setting the transaction isolation level to `ReadUncommitted` but it may not fully solve your use case due to the nature of the operations you're performing.

The `ReadUncommitted` isolation level allows a transaction to see uncommitted changes from other transactions, which is also known as a `dirty read`. However, this behavior can lead to various anomalies and is not generally recommended.

In your specific use case, you're trying to perform multiple operations on the same group in a single transaction. Each operation reads the group info and then tries to update it. If two operations try to update the same group at the same time, they could potentially see inconsistent data due to the `ReadUncommitted` isolation level.

One possible solution could be to refactor your code to perform all operations on a single group within a single transaction. This would ensure that all operations on a group are performed sequentially and that they see a consistent view of the group data.
Here's an example:
```js
async function doThingWhereINeedToRemoveUsersFromTheirGroup() {
  const groups = {} // a map to store actions for each group

  for (let user of users) {
    const removeActions = await removeFromGroup(user.id)
    const groupId = removeActions[0].group.id // assuming the group id is available here

    if (!groups[groupId]) {
      groups[groupId] = []
    }

    groups[groupId].push(...removeActions)
  }

  // perform all operations on each group in a single transaction
  for (let groupId in groups) {
    await prisma.$transaction(groups[groupId], {
      isolationLevel: Prisma.TransactionIsolationLevel.ReadUncommitted,
    })
  }
}
```
In this example, all operations on a single group are performed within a single transaction, which should prevent the issue of concurrent updates to the same group.

Another possible solution is to implement `Optimisitic Locking` using a versioning column on the `groupMember` table, like version. During `removeFromGroup`, read the version of the `groupMember` alongside other details. In the update query, include `where: { id: userId, version: groupMember.version }`. This ensures the update succeeds only if the specific version of the `groupMember` hasn't changed since the initial read.

If a concurrent transaction removes the same user before your update, the Prisma client will throw a `PrismaClientKnownRequestError`. You can then handle this error by retrying the removal logic within a reasonable limit.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by sudall** [reply]
Hi, I don't want to use `ReadUncommitted`. I'm also already using optimistic concurrency on the `Group` via `membersVersion`, which is where I'm encountering the issue. I want to be able to compose interactive transaction pieces together so that they can see each other's uncommitted changes OR somehow be able to see uncommitted changes for a transaction outside of an interactive transaction.

**Reply by ludralph** [reply]
Hi @sudall 

Prisma does offer an interactive transaction API, but it doesn't support composing multiple interactive transactions together in the way you've described.  It doesn't provide a built-in way to see uncommitted changes from one transaction in another transaction or outside of an interactive transaction. This is related to the isolation levels of the underlying database, which Prisma doesn't expose direct control over in its API. 

As a workaround,  you could use raw SQL queries within a transaction to set the desired isolation level and perform the operations.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
